---
title: "Introduction to SNPannotator package"
subtitle: "Investigating the Functional Characteristics of Selected SNPs and Their Vicinity Genomic Region"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to SNPannotator package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


## Overview

SNPannotator is a bioinformatics tool designed to annotate genetic variants detected in genome-wide association (GWAS) studies. While GWAS identifies statistical associations between genetic variants and phenotypic traits, it does not explain the biological mechanisms underlying these associations. Post-GWAS analysis is therefore essential to fill this gap by determining the functional impact of these variants on human biology. This manual provides a step-by-step guide to using the SNPannotator package for post-GWAS analysis, helping researchers better understand the genetic architecture of complex traits and diseases.

---


## Installation

```{r, echo = FALSE}
  knitr::opts_chunk$set(
  eval=FALSE,
  results = "hide",
  collapse = TRUE, 
  comment = "#>",
  results = "asis"
)
```

The easiest way to install the SNPannotator package is to get it from the Comprehensive R Archive Network (CRAN). Required dependencies will be downloaded automatically:

```{r}
    # this will automatically download and install the dependencies.
    install.packages("SNPannotator")
```

---

## Input Preparation

SNPannotator requires two input files:

1. GWAS top hit variants: 

A text file containing a list of the top independent GWAS variants. Each line of this file corresponds to a single variant rsID, which represents a genetic locus associated with the studied trait.


2. Configuration File: 

A text-based configuration file that users must edit before running the analysis. The configuration file can be obtained using the following command:

```{r}
# first, load the library
Library(SNPannotator)

# save the configuration file to a specific folder
getConfigFile('/home/user/project1/postGWAS')
```

---

## Step by step guide to running the package

1- Create an empty folder to save the output result files. Ensure that current user and R environment has the necessary read/write permissions for this folder.

2- Prepare the variants input file. Each line of the file should correspond to a single variant rsID. It is recommended to place this file in the newly created folder for easy access.

3- Obtain the configuration file from the package using the command `getConfigFile('/path/to/folder')`. It is recommended to copy this file into the newly created folder for easy access. Edit the parameters of this file as needed.

4- Run the annotation pipeline by running `annotate('/path/to/config.ini')`.

---

## Annotation steps of the top GWAS variants

The steps included in this pipeline are:

- Identifying variants in linkage disequilibrium (LD) with top GWAS hits

- Assessing variant information and their regulatory potential

- Evaluating their impact on gene expression through eQTL analysis

- Conducting enrichment analysis to understand broader biological networks

- Finding the previously found associated traits


---

## Package functions:


### getConfigFile() 

Copy a sample configuration file to specified folder.

Function parameters:

-	dir.path, The existing folder for copying the file.

Example:

```{r}

library(SNPannotator)

getConfigFile('/home/user/config.ini')

```

---

### annotate() 

This is the main package function, which receives the path to a configuration file (.ini) for running the parameter.

Function parameters:

-	configurationFilePath, The path to the configuration file.

-	verbose, Whether to display messages in the console.


Example:

```{r}

annotate('/home/user/analysis/config.ini')
```

---

### findProxy() 

this function can be used to find variants that are in high LD with a list of selected variants. 


Function parameters:

- rslist, A vector of rs numbers.

- file, Path to the Excel file for saving search results. This is optional, the result will be just available in R environment if file path is not provided.

- build, Genome build. Either 37 or 38. default: 37

- db, The population database for calculating LD scores.

- window_size, Number of base pairs around the variant for checking LD scores (max = 500kb).

- r2, The minimum LD threshold for selecting variants around the target SNP. default: 0.8. 

Example:

```{r}
FindProxy("rs234")

FindProxy(c("rs234","rs678"))
```

---

### pairwiseLD() 

This function computes the linkage disequilibrium (LD) between the selected variants using data from the Ensembl website.

Function parameters

-	rslist, A vector of rs numbers.

-	file, Path to the Excel file for saving search results. This is optional, the result will be just available in R environment if file path is not provided.

-	pairwise, If TRUE, compute pairwise LD between all elements of a list. If FALSE, computes the LD between first and other elements of the list. default: FALSE

-	build, Genome build. Either 37 or 38. default: 37

-	db, The population database for calculating LD scores.

-	window_size, Number of base pairs around the variant for checking LD scores (max = 500kb).

-	r2, The minimum LD threshold for selecting variants around the target SNP. default: 0.8. 

Example:

```{r}

output = pairwiseLD(c('rs234','rs10244533'))

```
